src/target/riscv/riscv.c

```diff
@@ -3099,22 +3117,41 @@ static int riscv_write_phys_memory(struct target *target, target_addr_t phys_add
 static int riscv_write_memory(struct target *target, target_addr_t address,
 		uint32_t size, uint32_t count, const uint8_t *buffer)
 {
+	uint32_t current_count = 0;
+
 	if (count == 0) {
 		LOG_TARGET_WARNING(target, "0-length write to 0x%" TARGET_PRIxADDR, address);
 		return ERROR_OK;
 	}
 
-	target_addr_t physical_addr;
-	int result = target->type->virt2phys(target, address, &physical_addr);
-	if (result != ERROR_OK) {
-		LOG_TARGET_ERROR(target, "Address translation failed.");
-		return result;
-	}
-
 	struct target_type *tt = get_target_type(target);
 	if (!tt)
 		return ERROR_FAIL;
-	return tt->write_memory(target, physical_addr, size, count, buffer);
+
+	if (address % size) {
+		LOG_TARGET_ERROR(target, "Address 0x%" TARGET_PRIxADDR " not aligned to size %d", address, size);
+		return ERROR_FAIL;
+	}
```

Comment:
AFAIU, misaligned access can be supported and is supported in current OpenOCD.

However, IMHO, accessing an element across a page boundary can be addressed later.

I'd suggest something like:

```suggestion
        const target_addr_t page_offset_mask = RISCV_PGSIZE - 1;
	const bool crosses_page_boundary = (address + size * count - 1) & ~page_offset_mask != address & ~page_offset_mask;
	if (address % size && crosses_page_boundary) {
		int mmu_enabled;
		int res = riscv_mmu(target, &mmu_enabled);
		if (res != ERROR_OK)
			return res;
		if (mmu_enabled)
			LOG_TARGET_ERROR(target, "Accessing an element across a page boundary is not supported.");
		return ERROR_FAIL;
	}
```

Reply:
Thanks. Addressed.